adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1148,14 +1149,92 @@ protected final void onCommandReceived(
         final String commandRequestId = (command.isOneWay() ? "" : command.getRequestId());
         final String topic = String.format("%s/%s/%s/%s/%s/%s", subscription.getEndpoint(), tenantId, deviceId,
                 subscription.getRequestPart(), commandRequestId, command.getName());
-        endpoint.publish(topic, command.getPayload(), qos, false, false);
+        LOG.debug("Publishing command to device [tenant-id: {}, device-id: {}, MQTT client-id: {}, QoS: {}]",
+                        subscription.getTenant(), subscription.getDeviceId(), subscription.getClientId(), subscription.getQos());
+        commandContext.getCurrentSpan().log(String.format("Publishing command to device [tenant-id: %s, device-id: %s, MQTT client-id: %s, QoS: %s]",
+                subscription.getTenant(), subscription.getDeviceId(), subscription.getClientId(), subscription.getQos()));
+
+        if (MqttQoS.AT_LEAST_ONCE.equals(subscription.getQos())) {
+            registerPublishAcknowledgeHandler(endpoint, subscription);
+        }
+
+        endpoint.publish(topic, command.getPayload(), subscription.getQos(), false, false, publishSentHandler -> {
+            if (publishSentHandler.succeeded()) {
+                if (MqttQoS.AT_LEAST_ONCE.equals(subscription.getQos())) {
+                    subscription.addToWaitingForAcknowledgement(publishSentHandler.result(),
+                            startAcknowledgementTimer(publishSentHandler.result(), subscription), commandContext);
+                } else {
+                    afterCommandPublished(subscription, commandContext);
+                }
+            } else {
+                LOG.error(
+                        "Error publishing command to device [tenant-id: {}, device-id: {}, MQTT client-id: {}, QoS: {}]",
+                        subscription.getTenant(), subscription.getDeviceId(), endpoint.clientIdentifier(),
+                        subscription.getQos(),
+                        publishSentHandler.cause());
+                TracingHelper.logError(commandContext.getCurrentSpan(), publishSentHandler.cause());
```

Comment:
IMHO the command needs to be released in this case, doesn't it?

Reply:
yes, it needs to be released. I have updated the code.